File: brainstem/command_sender.py
"""Command Sender — signs and sends commands to Vercel."""
import asyncio
import logging
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, Optional

import aiohttp

from brainstem.auth import BrainstemAuth
from brainstem.config import BrainstemConfig

logger = logging.getLogger("jarvis.brainstem.sender")


class CommandSender:
    def __init__(self, config: BrainstemConfig, auth: BrainstemAuth) -> None:
        self._config = config
        self._auth = auth
        self._session: Optional[aiohttp.ClientSession] = None

    async def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession()
        return self._session

    def build_payload(
        self,
        text: str,
        priority: str = "realtime",
        response_mode: str = "stream",
        intent_hint: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        payload: Dict[str, Any] = {
            "command_id": str(uuid.uuid4()),
            "device_id": self._config.device_id,
            "device_type": self._config.device_type,
            "text": text,
            "priority": priority,
            "response_mode": response_mode,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        if intent_hint:
            payload["intent_hint"] = intent_hint
        if context:
            payload["context"] = context
        payload["signature"] = self._auth.sign(payload)
        return payload

    async def send_command(
        self,
        text: str,
        priority: str = "realtime",
        response_mode: str = "stream",
        intent_hint: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        payload = self.build_payload(
            text=text,
            priority=priority,
            response_mode=response_mode,
            intent_hint=intent_hint,
            context=context,
        )
        session = await self._get_session()
        url = f"{self._config.vercel_url}/api/command"
        logger.info("[Sender] %s -> %s", text[:50], url)
        async with session.post(url, json=payload) as resp:
            content_type = resp.headers.get("content-type", "")
            if "text/event-stream" in content_type:
                return {"status": "streaming", "command_id": payload["command_id"]}
            else:
                return await resp.json()

    async def get_frontmost_app(self) -> str:
        try:
            proc = await asyncio.create_subprocess_exec(
                "osascript", "-e",
                'tell application "System Events" to get name of first application process whose frontmost is true',
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.DEVNULL,
            )
            stdout, _ = await proc.communicate()
            return stdout.decode().strip() if stdout else "unknown"
        except Exception:
            return "unknown"

    async def close(self) -> None:
        if self._session:
            await self._session.close()

File: JARVIS-Apple/JARVISKit/Sources/JARVISKit/Networking/CommandSender.swift
import Foundation

public final class CommandSender: Sendable {
    private let baseURL: String
    private let auth: DeviceAuth

    public init(baseURL: String, auth: DeviceAuth) {
        self.baseURL = baseURL
        self.auth = auth
    }

    public func send(_ text: String, priority: Priority = .realtime, responseMode: ResponseMode = .stream, intentHint: String? = nil, context: CommandContext? = nil) async throws -> CommandResponse {
        var payload = CommandPayload(
            deviceId: auth.deviceId,
            deviceType: auth.deviceType,
            text: text,
            intentHint: intentHint,
            context: context,
            priority: priority,
            responseMode: responseMode
        )
        payload.signature = auth.sign(payload)

        let url = URL(string: "\(baseURL)/api/command")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let encoder = JSONEncoder()
        request.httpBody = try encoder.encode(payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        let httpResponse = response as! HTTPURLResponse

        if httpResponse.statusCode == 401 {
            throw JARVISError.authFailed
        }

        let contentType = httpResponse.value(forHTTPHeaderField: "Content-Type") ?? ""
        if contentType.contains("text/event-stream") {
            return CommandResponse(status: "streaming", commandId: payload.commandId)
        }

        return try JSONDecoder().decode(CommandResponse.self, from: data)
    }
}

public struct CommandResponse: Codable, Sendable {
    public let status: String
    public let commandId: String?
    public let jobId: String?
    public let brain: String?

    public init(status: String, commandId: String? = nil, jobId: String? = nil, brain: String? = nil) {
        self.status = status
        self.commandId = commandId
        self.jobId = jobId
        self.brain = brain
    }

    enum CodingKeys: String, CodingKey {
        case status
        case commandId = "command_id"
        case jobId = "job_id"
        case brain
    }
}
